tensorflow_similarity/losses/simclr.py

```diff
@@ -16,81 +16,122 @@
     A Simple Framework for Contrastive Learning of Visual Representations
     https://arxiv.org/abs/2002.05709
 """
-from typing import Any, Callable, Dict, Optional
+from typing import Any, Dict
 
 import tensorflow as tf
 
 from tensorflow_similarity.types import FloatTensor
 
-LARGE_NUM = 1e9
-
 
 @tf.keras.utils.register_keras_serializable(package="Similarity")
 class SimCLRLoss(tf.keras.losses.Loss):
     """SimCLR Loss.
     [A Simple Framework for Contrastive Learning of Visual Representations](https://arxiv.org/abs/2002.05709)
-    code adapted from [orignal github](https://github.com/google-research/simclr/tree/master/tf2)
+    code adapted from [original github](https://github.com/google-research/simclr/tree/master/tf2)
     """
 
-    def __init__(
-        self,
-        temperature: float = 0.05,
-        margin: float = 0.001,
-        reduction: Callable = tf.keras.losses.Reduction.AUTO,
-        name: Optional[str] = None,
-        **kwargs
-    ):
-        super().__init__(reduction=reduction, name=name, **kwargs)
+    LARGE_NUM = 1e9
+
+    def __init__(self, temperature: float = 0.05, **kwargs):
+        super().__init__(**kwargs)
         self.temperature = temperature
-        self.margin = margin
+
+    def contrast(self, hidden1: FloatTensor, hidden2: FloatTensor) -> FloatTensor:
+
+        # local replica batch size
+        batch_size = tf.shape(hidden1)[0]
+
+        if not tf.distribute.in_cross_replica_context():
+            # SimCLR loss computes similarity with 2N-1 other examples, when N is the global_batch_size.
+            # In distributed training, each replica sees only N/n_replicas examples and to compare with all 2N-1
+            # examples we must first get them from other replicas.
+            strategy = tf.distribute.get_strategy()
+            hidden1_large = SimCLRLoss.cross_replica_concat(hidden1, strategy)
+            hidden2_large = SimCLRLoss.cross_replica_concat(hidden2, strategy)
+            enlarged_batch_size = tf.shape(hidden1_large)[0]
+            replica_context = tf.distribute.get_replica_context()
+            labels_idx = tf.range(batch_size) + replica_context.replica_id_in_sync_group * batch_size
+            labels = tf.one_hot(labels_idx, enlarged_batch_size * 2)
+            masks = tf.one_hot(labels_idx, enlarged_batch_size)
+        else:
+            hidden1_large = hidden1
+            hidden2_large = hidden2
+            labels = tf.one_hot(tf.range(batch_size), batch_size * 2)
+            masks = tf.one_hot(tf.range(batch_size), batch_size)
+
+        # compute pairwise
+        ab = tf.matmul(hidden1, hidden2_large, transpose_b=True) / self.temperature
+        aa = tf.matmul(hidden1, hidden1_large, transpose_b=True) / self.temperature
```

Comment:
Should this be hidden2?

`tf.matmul(hidden2, hidden1_large, transpose_b=True)`

Reply:
no. aa holds the cosine similarity between all vectors in hidden1 and itself. later we make the diag -inf to remove the dist(e_i,e_i) from the denominator in thesum :
<img width="337" alt="image" src="https://user-images.githubusercontent.com/4004127/173762038-672ffedf-2a0a-4887-9495-8eaf10f57e52.png">
Its like it was before:
`aa = tf.matmul(za, za, transpose_b=True)`

Reply:
You're totally right, I'd forgotten about that. Thanks for clarifying and for the PR.